_posts/2018-12-21-why-is-swiftpm-taking-so-long.md

```diff
@@ -0,0 +1,145 @@
+is the underlying tooling **above** the compiler but **below** Xcode: `xcodebuild`, wouldn't make any sense in that
+context. It's deeply macOS and Cocoa specific.
+
+So, Apple needed something to handle the following problems if you were working in a Xcode-less Swift project:
+
+- Handling build dependencies
+- Linking to existing code
+- Creating library abstractions
+- Creating executable for the language
+- Handling sharing code between projects
+
+Roughly in that order of precedence too. In short, they needed a build system first, and then a dependency manager
+to support remote packages.
+
+There are a few big existing build systems: [Buck][buck] and [Ninja][ninja] are the biggest players in the game,
+but they aim to be language agnostic build tools. This makes them a bit too abstract to be useful as the main way
+to build a modular system for building projects in a language. So, each language from Swift's generation of _"As
+fast as C, but safe"_ like Rust, Go and D all implement a build system unique to their language.
+
+# llbuild
+
+Apple's approach to building a build system is smart. Instead of from scratch creating a build tool for Swift, they
+created an abstraction which makes it feasible to build complex builds systems quicker. That's [llbuild][], a
+low-level build system. This was used to create the Swift build system. Then, llbuild was re-used to create
+[Xcode's build system][xcbuild].
+
+The new build system for Xcode came out in 2017 with Xcode 9, and became the default in 2018 with Xcode 10. What's
+impressive about the new build system is how much it was just like the last, but better, to quote [James
+Dempsey][jd-build-system]:
+
+> New build system. Same build settings
+
+Which is impressive, because the existing system is [complex][bs].
+
+# Build System != Package Manager
+
+CocoaPods was "useful" for Cocoa development very quickly in comparison (roughly a month from initial commit to
+supporting a few existing libraries in an app back in 2011), this was because CocoaPods completely skipped the
+build tool stage. There was no need, Xcode has a working build system and you can control it via `.xcproject`
+files. This meant that the key output for CocoaPods is to reliably support generating xcode projects and the
+underlying abstractions making it actually compile is left to the Xcode team.
+
+The Swift PM team couldn't do this without making it deeply dependent on Xcode. CocoaPods can be indifferent to
+something like Linux support (fun fact: last month [Windows support][winders] was added) because no-one is writing
+apps on Linux, and it's only concern is shipping those Xcode project files.
+
+This let CocoaPods to work on comprehensive support of the existing library ecosystem, rather than the more mundane
+task of actually figuring out how to make a project compile. Meaning CocoaPods can focus on more user-facing issues
```

Comment:
I wouldn’t say this is entirely correct, we still had to do plenty of figuring out how to get projects to compile and deal directly with compiler/linker options. However, the major difference may be that many of these options have long been discussed and understood in the regular C community? Anyways, suffice to say that we definitely didn’t need to do the amount of work that the Swift PM team needs to do, such as e.g. determining how products can even be build, instead of us relying on the few options that Apple gave us.

Reply:
https://github.com/artsy/artsy.github.io/pull/511/commits/0d69bf4b7f6e97f3aa4bed224162b310216fc176